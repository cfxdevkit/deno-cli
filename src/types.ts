export interface Block {
	blockNumber?: bigint
	timestamp: bigint | number
	number?: bigint // For espace block
}

export interface BlockInfo {
	number: bigint
	timestamp: string
	label: string
}

export interface Transaction {
	hash: string
	from: string
	to: `0x${string}` | null
	value: bigint | string
	gas: bigint
	gasPrice?: bigint | undefined
}

// --- Types ---
export type KeystoreEntry = {
	type: 'plaintext' | 'encoded'
	label: string
	mnemonic: string
}

export type KeystoreFile = {
	keystore: KeystoreEntry[]
	activeIndex: number | null
}
