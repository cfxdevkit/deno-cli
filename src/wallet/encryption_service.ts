import { crypto } from '@std/crypto'
import { Secret } from 'cliffy/prompt'

export class EncryptionService {
	async deriveKeyFromPassword(reason: string, salt: Uint8Array): Promise<CryptoKey> {
		const password = await Secret.prompt({ message: reason })
		const encoder = new TextEncoder()
		const passwordKey = await crypto.subtle.importKey(
			'raw',
			encoder.encode(password),
			'PBKDF2',
			false,
			['deriveKey'],
		)

		return crypto.subtle.deriveKey(
			{
				name: 'PBKDF2',
				salt,
				iterations: 100_000,
				hash: 'SHA-256',
			},
			passwordKey,
			{ name: 'AES-GCM', length: 256 },
			true,
			['encrypt', 'decrypt'],
		)
	}

	getRandomValues(array: Uint8Array) {
		return crypto.getRandomValues(array)
	}

	async encryptMnemonic(mnemonic: string): Promise<string> {
		const iv = this.getRandomValues(new Uint8Array(12))
		const salt = this.getRandomValues(new Uint8Array(16))
		const key = await this.deriveKeyFromPassword('Enter encryption password to secure your mnemonic.', salt)
		const encrypted = await crypto.subtle.encrypt(
			{ name: 'AES-GCM', iv },
			key,
			new TextEncoder().encode(mnemonic),
		)
		const encryptedData = new Uint8Array([...salt, ...iv, ...new Uint8Array(encrypted)])
		return btoa(String.fromCharCode(...encryptedData))
	}

	async decryptMnemonic(encryptedMnemonic: string): Promise<string> {
		const encryptedBytes = Uint8Array.from(atob(encryptedMnemonic), (c) => c.charCodeAt(0))
		const salt = encryptedBytes.slice(0, 16)
		const iv = encryptedBytes.slice(16, 28)
		const data = encryptedBytes.slice(28)

		for (let attempt = 1; attempt <= 3; attempt++) {
			try {
				const key = await this.deriveKeyFromPassword('Password: ', salt)
				const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, data)
				return new TextDecoder().decode(decrypted)
			} catch {
				console.error(`Decryption failed (${attempt}/3). Incorrect password or corrupted data.`)
			}
		}
		throw new Error('Maximum decryption attempts reached.')
	}
}
