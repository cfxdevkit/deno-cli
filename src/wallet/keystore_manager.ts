import { ensureFileSync } from '@std/fs'
import { join } from '@std/path'
import { KeystoreEntry, KeystoreFile } from '../types.ts'

export interface IKeystoreManager {
	readKeystore(): Promise<KeystoreFile | null>
	writeKeystore(): Promise<void>
	getKeystore(): KeystoreEntry[]
	setKeystore(keystore: KeystoreEntry[]): void
	getActiveIndex(): number
	setActiveIndex(index: number | null): void
}

export class KeystoreManager implements IKeystoreManager {
	public keystorePath: string
	public keystore: KeystoreEntry[] = []
	public activeIndex: number | null = 0

	constructor() {
		this.keystorePath = join(Deno.env.get('HOME') || '', '.devkit.keystore.json')
		ensureFileSync(this.keystorePath)
	}

	async readKeystore(): Promise<KeystoreFile | null> {
		try {
			const data = await Deno.readTextFile(this.keystorePath)
			return data.trim() ? JSON.parse(data) : null
		} catch (error) {
			if (error instanceof Deno.errors.NotFound) return null
			throw error
		}
	}

	async writeKeystore(): Promise<void> {
		const data: KeystoreFile = { keystore: this.keystore, activeIndex: this.activeIndex }
		await Deno.writeTextFile(this.keystorePath, JSON.stringify(data, null, 2))
	}

	getKeystore() {
		return this.keystore
	}

	setKeystore(keystore: KeystoreEntry[]) {
		this.keystore = keystore
	}

	setKeystorePath(keystore_path: string) {
		this.keystorePath = keystore_path
	}

	setActiveIndex(index: number | null) {
		this.activeIndex = index
	}

	getActiveIndex(): number {
		return this.activeIndex || 0
	}
}
