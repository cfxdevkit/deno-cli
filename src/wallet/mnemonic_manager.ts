import { Input, Select } from 'cliffy/prompt'
import { english } from 'viem/accounts'
import { KeystoreManager } from './keystore_manager.ts'
import { EncryptionService } from './encryption_service.ts'
import { HDWallet } from '@conflux-dev/hdwallet'

export class MnemonicManager {
	private keystoreManager: KeystoreManager
	public encryptionService: EncryptionService

	constructor(keystoreManager: KeystoreManager, encryptionService: EncryptionService) {
		this.keystoreManager = keystoreManager
		this.encryptionService = encryptionService
	}

	async addMnemonic(): Promise<void> {
		const storageChoice = await Select.prompt({
			message: 'Choose storage option for the mnemonic:',
			options: [
				{ name: 'Store encrypted', value: 'e' },
				{ name: 'Store in plaintext', value: 'p' },
			],
		})

		const mnemonic = await this.promptForMnemonic()

		const defaultLabel = `Mnemonic ${this.keystoreManager.getKeystore().length + 1}`
		const label = await Input.prompt({
			message: 'Enter a label for this mnemonic:',
			default: defaultLabel,
		})

		if (storageChoice === 'p') {
			this.keystoreManager.getKeystore().push({ type: 'plaintext', label, mnemonic })
		} else {
			const encryptedMnemonic = await this.encryptionService.encryptMnemonic(mnemonic)
			this.keystoreManager.getKeystore().push({ type: 'encoded', label, mnemonic: encryptedMnemonic })
		}

		await this.keystoreManager.writeKeystore()
		console.error(storageChoice === 'p' ? 'Mnemonic stored in plaintext.' : 'Mnemonic stored securely.')
	}

	private async promptForMnemonic(): Promise<string> {
		const userChoice = await Select.prompt({
			message: 'Generate or import a mnemonic?',
			options: [
				{ name: 'Generate a new mnemonic', value: 'g' },
				{ name: 'Insert an existing mnemonic', value: 'i' },
			],
		})
		return userChoice === 'i' ? await this.importMnemonic() : this.generateMnemonic()
	}

	generateMnemonic(): string {
		return HDWallet.generateMnemonic()
	}

	private async importMnemonic(): Promise<string> {
		console.log('\nPlease enter your mnemonic key one word at a time.')
		const words: string[] = []
		for (let i = 1; i <= 12; i++) {
			const word = await Input.prompt({
				message: `Enter word ${i} of 12`,
				suggestions: english,
				validate: (input) =>
					english.includes(input) || 'Invalid word. Please enter a valid BIP-39 mnemonic word.',
			})
			words.push(word)
		}
		return words.join(' ')
	}
}
