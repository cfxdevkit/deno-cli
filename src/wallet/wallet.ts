import { HDWallet } from '@conflux-dev/hdwallet'
import { KeystoreManager } from './keystore_manager.ts'
import { MnemonicManager } from './mnemonic_manager.ts'
import { Select } from 'cliffy/prompt'
import { generatePrivateKey } from 'viem/accounts'
import { EncryptionService } from './encryption_service.ts'

export class Wallet {
	private keystoreManager: KeystoreManager
	private mnemonicManager: MnemonicManager
	private hdWallet?: HDWallet

	constructor() {
		this.keystoreManager = new KeystoreManager()
		this.mnemonicManager = new MnemonicManager(this.keystoreManager, new EncryptionService())
	}

	async initializeKeystore(): Promise<void> {
		const existingKeystore = await this.keystoreManager.readKeystore()
		if (!existingKeystore) {
			console.log('No keystore found. Creating a default keystore...')
			this.keystoreManager.getKeystore().push({
				type: 'plaintext',
				label: 'Default Keystore',
				mnemonic: 'test test test test test test test test test test test junk',
			})
			this.keystoreManager.setActiveIndex(0)
			await this.keystoreManager.writeKeystore()
			console.log('Default keystore created and activated.')
		} else {
			this.keystoreManager.setKeystore(existingKeystore.keystore)
			this.keystoreManager.setActiveIndex(existingKeystore.activeIndex)
		}
		this.hdWallet = new HDWallet(await this.getActiveMnemonic())
	}

	async getActiveMnemonic(): Promise<string> {
		if (this.keystoreManager.getActiveIndex() === null || this.keystoreManager.getActiveIndex() < 0) {
			throw new Error('No active mnemonic selected.')
		}
		let mnemonic: string = ''
		if (!this.hdWallet) {
			const mnemonicObj = this.keystoreManager.getKeystore()[this.keystoreManager.getActiveIndex()]
			mnemonic = mnemonicObj.type === 'encoded'
				? await this.mnemonicManager.encryptionService.decryptMnemonic(mnemonicObj.mnemonic)
				: mnemonicObj.mnemonic
		} else {
			mnemonic = this.hdWallet.mnemonic
		}
		return mnemonic
	}

	async addMnemonic(): Promise<void> {
		await this.mnemonicManager.addMnemonic()
	}

	async selectActiveMnemonic(): Promise<void> {
		const currentIndex = this.keystoreManager.getActiveIndex()
		const selectedIndex = await Select.prompt({
			message: 'Select the active mnemonic:',
			options: this.keystoreManager.getKeystore().map((mnemonicObj, index) => ({
				name: mnemonicObj.label,
				value: String(index),
			})),
		})

		this.keystoreManager.setActiveIndex(Number(selectedIndex))
		await this.keystoreManager.writeKeystore()
		if (Number(selectedIndex.value) !== currentIndex) {
			this.hdWallet = new HDWallet(await this.getActiveMnemonic())
		}
		console.log(
			`Active wallet set to: ${this.getActiveMnemonicLabel()}`,
		)
	}

	generatePrivateKey(): `0x${string}` {
		return generatePrivateKey()
	}

	getActiveMnemonicLabel(): string {
		if (this.keystoreManager.getActiveIndex() === null || this.keystoreManager.getActiveIndex() < 0) {
			throw new Error('No active mnemonic selected.')
		}
		const mnemonicObj = this.keystoreManager.getKeystore()[this.keystoreManager.getActiveIndex()]
		return mnemonicObj.label
	}

	privateKeyByDerivationPath(derivationPath: string): string {
		return `0x${this.hdWallet!.getPrivateKey(derivationPath).toString('hex')}`
	}

	espacePrivateKey(index: number): string {
		return this.privateKeyByDerivationPath(`m/44'/60'/0'/0/${index}`)
	}

	corePrivateKey(index: number): string {
		return this.privateKeyByDerivationPath(`m/44'/503'/0'/0/${index}`)
	}
}
