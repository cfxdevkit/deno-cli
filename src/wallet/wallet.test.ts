import { assertEquals } from '@std/assert'
import { Wallet } from './wallet.ts'
import { join } from '@std/path'
import { KeystoreManager } from './keystore_manager.ts'
import { KeystoreEntry } from '../types.ts'
const mnemonic = 'test test test test test test test test test test test junk'

Deno.test('KeystoreManager - initialize and read keystore', async () => {
	const tmpdir = await Deno.makeTempDir()
	const tempKeystorePath = join(tmpdir, '.devkit.keystore.json')
	// Mock Deno.env to set HOME to current directory
	const originalEnv = Deno.env.get
	Deno.env.get = (key: string) => (key === 'HOME' ? tmpdir : originalEnv(key))

	try {
		const wallet = new Wallet()
		await wallet.initializeKeystore()

		assertEquals(await wallet.getActiveMnemonic(), mnemonic)
		assertEquals(wallet.getActiveMnemonicLabel(), 'Default Keystore')
	} finally {
		// Clean up test file and restore environment
		await Deno.remove(tempKeystorePath)
		Deno.env.get = originalEnv
	}
})

Deno.test('KeystoreManager - Initialize keystore when keystore exists', async () => {
	const tmpdir = await Deno.makeTempDir()
	const tempKeystorePath = join(tmpdir, '.devkit.keystore.json')
	// Mock Deno.env to set HOME to current directory
	const originalEnv = Deno.env.get
	Deno.env.get = (key: string) => (key === 'HOME' ? tmpdir : originalEnv(key))

	try {
		// await wallet.initializeKeystore();
		const km = new KeystoreManager()
		const existingKeystore = {
			keystore: [
				{ type: 'plaintext', label: 'My Mnemonic', mnemonic: 'test test test' } as KeystoreEntry,
			],
			activeIndex: 0,
		}

		km.setKeystore(existingKeystore.keystore)
		km.setKeystorePath(tempKeystorePath)
		km.setActiveIndex(existingKeystore.activeIndex)
		km.writeKeystore()
		km.readKeystore()
		const wallet = new Wallet()
		await wallet.initializeKeystore()

		const activeMnemonic = await wallet.getActiveMnemonic()
		assertEquals(activeMnemonic, 'test test test')
	} finally {
		// Clean up test file and restore environment
		await Deno.remove(tempKeystorePath)
		Deno.env.get = originalEnv
	}
})

// // // Test: Add a mnemonic
// // await snapshotTest({
// //     name: "Wallet - Add mnemonic",
// //     meta: import.meta,
// //     stdin: ansi.text("\n") // Simulate prompts for adding a mnemonic
// //         .cursorDown()
// //         .text("\n") // Choose "Store plaintext"
// //         .toArray(),
// //     async fn() {
// //         mockKeystoreManager.setKeystore([]);
// //         mockKeystoreManager.setActiveIndex(null);

// //         await wallet.addMnemonic();

// //         const keystore = mockKeystoreManager.getKeystore();
// //         assertEquals(keystore.length, 1);
// //         assertEquals(keystore[0].label, "Mnemonic 1");
// //     },
// // });

// // // Test: Select active mnemonic
// // await snapshotTest({
// //     name: "Wallet - Select active mnemonic",
// //     meta: import.meta,
// //     stdin: ansi
// //         .cursorDown()
// //         .text("\n") // Select second mnemonic in the list
// //         .toArray(),
// //     async fn() {
// //         mockKeystoreManager.setKeystore([
// //             { type: 'plaintext', label: 'Mnemonic 1', mnemonic: 'mnemonic 1' },
// //             { type: 'plaintext', label: 'Mnemonic 2', mnemonic: 'mnemonic 2' },
// //         ]);
// //         mockKeystoreManager.setActiveIndex(0);

// //         await wallet.selectActiveMnemonic();

// //         assertEquals(mockKeystoreManager.getActiveIndex(), 1);
// //     },
// // });

// // // Test: Generate private key
// // await snapshotTest({
// //     name: "Wallet - Generate private key",
// //     meta: import.meta,
// //     stdin: ansi.toArray(),
// //     async fn() {
// //         const privateKey = wallet.generatePrivateKey();
// //         assertEquals(privateKey.startsWith("0x"), true);
// //         assertEquals(privateKey.length, 66); // 64 hex characters + "0x"
// //     },
// // });

// // // Test: Retrieve active mnemonic label
// // await snapshotTest({
// //     name: "Wallet - Get active mnemonic label",
// //     meta: import.meta,
// //     stdin: ansi.toArray(),
// //     async fn() {
// //         mockKeystoreManager.setKeystore([
// //             { type: 'plaintext', label: 'Mnemonic 1', mnemonic: 'test mnemonic' },
// //         ]);
// //         mockKeystoreManager.setActiveIndex(0);

// //         const label = wallet.getActiveMnemonicLabel();
// //         assertEquals(label, 'Mnemonic 1');
// //     },
// // });

// // // Test: Generate private key by derivation path
// // await snapshotTest({
// //     name: "Wallet - Generate private key by derivation path",
// //     meta: import.meta,
// //     stdin: ansi.toArray(),
// //     async fn() {
// //         mockKeystoreManager.setKeystore([
// //             { type: 'plaintext', label: 'Mnemonic 1', mnemonic: 'test test test' },
// //         ]);
// //         mockKeystoreManager.setActiveIndex(0);

// //         await wallet.initializeKeystore();
// //         const privateKey = wallet.privateKeyByDerivationPath("m/44'/60'/0'/0/0");
// //         assertEquals(privateKey.startsWith("0x"), true);
// //     },
// // });
