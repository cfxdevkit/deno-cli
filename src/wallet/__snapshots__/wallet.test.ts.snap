export const snapshot = {};

snapshot[`Wallet - Initialize keystore (no existing keystore) 1`] = `
stdout:
""
stderr:
""
`;
